src/Storages/MaterializedView/RefreshTask.cpp

```diff
@@ -1008,41 +1009,92 @@ void RefreshTask::interruptExecution()
 
 std::tuple<StoragePtr, TableLockHolder> RefreshTask::getAndLockTargetTable(const StorageID & storage_id, const ContextPtr & context)
 {
-    StoragePtr storage;
-    TableLockHolder storage_lock;
+    ///  1. Get table by name.
+    ///  2. Check that it's not dropped locally.
+    ///     (After that, it can't be dropped during the query because we're holding StoragePtr and
+    ///      TableLockHolder.)
+    ///     If this fails, retry and expect to see a different table by the same name.
+    ///  3. Do SYSTEM SYNC REPLICA. May fail if the table is being dropped.
+    ///     If this fails, retry until we see a different table by the same name.
+
+    StoragePtr prev_storage;
+    /// (Without maybe_unused, clang-tidy-19 seems to produce incorrect error
+    ///  "Value stored to 'prev_table_dropped_locally' is never read" about the assignment
+    ///  prev_table_dropped_locally = false further down.)
+    [[maybe_unused]] bool prev_table_dropped_locally = false;
+    std::exception_ptr exception;
+
     for (int attempt = 0; attempt < 10; ++attempt)
     {
-        StoragePtr prev_storage = std::move(storage);
-        storage = DatabaseCatalog::instance().getTable(storage_id, context);
+        if (attempt > 0 && !prev_table_dropped_locally)
+        {
+            /// We're waiting for DatabaseReplicated to catch up and see the new table.
+            std::this_thread::sleep_for(std::chrono::milliseconds(100));
+        }
+
+        StoragePtr storage = DatabaseCatalog::instance().getTable(storage_id, context);
```

Comment:
And this cannot fail under normal circumstances (when only RMV modifies the target) because of RENAME with exchange?

Reply:
Yes. `EXCHANGE TABLES` is not very atomic (https://github.com/ClickHouse/ClickHouse/issues/67646), but at least the `DatabaseCatalog` part of it is atomic.